From 24ebca9495e8d9f8af9b2eedfca5f46c6f178860 Mon Sep 17 00:00:00 2001
From: Patrik Jakobsson <patrik.r.jakobsson@gmail.com>
Date: Fri, 12 Jul 2013 15:38:52 +0200
Subject: drm/gma500: Add generic set_config() function

Signed-off-by: Patrik Jakobsson <patrik.r.jakobsson@gmail.com>
(cherry picked from commit 924cb5ffd81d66cc6461de955f7cb144cb3b7b6d)
Signed-off-by: Patrik Jakobsson <patrik.r.jakobsson@gmail.com>
---
 drivers/gpu/drm/gma500/gma_display.c |   16 ++++++++++++++++
 drivers/gpu/drm/gma500/gma_display.h |    3 +++
 2 files changed, 19 insertions(+)

--- a/drivers/gpu/drm/gma500/gma_display.c
+++ b/drivers/gpu/drm/gma500/gma_display.c
@@ -519,6 +519,22 @@ void gma_crtc_destroy(struct drm_crtc *c
 	kfree(psb_intel_crtc);
 }
 
+int gma_crtc_set_config(struct drm_mode_set *set)
+{
+	struct drm_device *dev = set->crtc->dev;
+	struct drm_psb_private *dev_priv = dev->dev_private;
+	int ret;
+
+	if (!dev_priv->rpm_enabled)
+		return drm_crtc_helper_set_config(set);
+
+	pm_runtime_forbid(&dev->pdev->dev);
+	ret = drm_crtc_helper_set_config(set);
+	pm_runtime_allow(&dev->pdev->dev);
+
+	return ret;
+}
+
 /**
  * Save HW states of given crtc
  */
--- a/drivers/gpu/drm/gma500/gma_display.h
+++ b/drivers/gpu/drm/gma500/gma_display.h
@@ -22,6 +22,8 @@
 #ifndef _GMA_DISPLAY_H_
 #define _GMA_DISPLAY_H_
 
+#include <linux/pm_runtime.h>
+
 struct gma_clock_t {
 	/* given values */
 	int n;
@@ -80,6 +82,7 @@ extern void gma_crtc_prepare(struct drm_
 extern void gma_crtc_commit(struct drm_crtc *crtc);
 extern void gma_crtc_disable(struct drm_crtc *crtc);
 extern void gma_crtc_destroy(struct drm_crtc *crtc);
+extern int gma_crtc_set_config(struct drm_mode_set *set);
 
 extern void gma_crtc_save(struct drm_crtc *crtc);
 extern void gma_crtc_restore(struct drm_crtc *crtc);
